### ISSUE TITLE
Detaching a distribution's `log_prob` to block gradients only w.r.t its parameters

### ISSUE THREAD

**Issue by iffsid** [pytorch/pytorch#25783]
## 🚀 Feature
It would be quite useful to have the general ability, to compute a 'detached' `log_prob` for any distribution; i.e. blocking all gradient computation w.r.t that distribution's parameters (but not the value), _without_ resorting to the alternative of constructing an instance of that distribution with explicitly detached parameters.

cc: @fritzo (a late request following our conversation at icml :slightly_smiling_face: )

## Motivation
This would be particularly useful for a variety of stochastic variational inference approaches [1,2], which seek to avoid the _score-function gradient_ component of the estimators, by either ignoring it entirely [1] or by _doubly reparameterising_ the estimator to include it in a form that avoids the larger variance of the score-function gradient estimator.

[1] [Sticking the Landing: Simple, Lower-VarianceGradient Estimators for Variational Inference](https://arxiv.org/abs/1703.09194)
[2] [Doubly Reparameterized Gradient Estimators for Monte Carlo Objectives](https://arxiv.org/abs/1810.04152)

## Pitch
The idea is to have an operation that allows a `log_prob` call to block gradient computation w.r.t the _parameters_, while gradients w.r.t the _value_ still get computed. Simply detaching the entire `log_prob` computation will not satisfy this requirement.

``` python
import torch
import torch.distributions as dist

params = torch.zeros(3,2), torch.ones(3,2)  # mu, sigma
pz = dist.Normal(*params)
z = pz.rsample()
# hypothetical setup
lpz = pz.log_prob(z, detached_params=True)  # or some similar form
loss = -lpz
```

## Alternatives

While one solution for this is to just create a new instance of the distribtuion in question with pre-detached parameters, for more complex models (as in those used with Pyro), this can get hairy quite quickly.

``` python
import torch
import torch.distributions as dist

params = torch.zeros(3,2), torch.ones(3,2)  # mu, sigma
params_ = tuple(p.detach() for p in params)
pz = dist.Normal(*params)
pz_ = dist.Normal(*params_)
z = pz.rsample()
lpz = pz_.log_prob(z)  # note: using pz_ to score
loss = -lpz
```

It also doesn't seem to be feasible to wrap the existing `log_prob` method, as there does not appear to be a way to access the all parameters of an arbitrary distribution---with a view to detaching them when required.

**Comment by vadimkantorov** [comment]
api idea: maybe intro a method `detach()` on distributions and call it `dist.detach().rsample(...)`. or is it too ambiguous?

**Comment by fritzo** [comment]
> there does not appear to be a way to access the all parameters of an arbitrary distribution

That's right, there is no generic way to access all parameters of a distribution object. The crucial blocker here is the early design decision to allow multiple different parametrizations in a single class, e.g. `Categorical` and `Bernoulli` can accept either `probs` or `logits`; `MultivariateNormal` can accept either `covariance_matrix`, `precision_matrix`, or `scale_tril`. One possible way forward is to split these into multiple classes dispatched by a wrapper function (as in [numpyro](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/distributions/discrete.py) and [funsor](https://github.com/pyro-ppl/funsor/blob/master/funsor/distributions.py)), but this would break inheritance patterns of much existing code.

> api idea: maybe intro a method `detach()`

I agree that's a reasonable api, but until we can generically access parameters, `.detach` would require bespoke implementations for each class, which seems prohibitively complex.

I think an easy workaround for now is to simply detach inputs; this is the same pattern used to change device or datatype (`.to()`).

**Comment by iffsid** [comment]
> I think an easy workaround for now is to simply detach inputs; this is the same pattern used to change device or datatype (`.to()`).

Do you mean like in the **Alternatives**? If so, would that not be an issue for usage, where the issue becomes needing two copies of every distribution instance---one to sample from, and one to score in?

**Comment by fritzo** [comment]
Yes, it's easiest to keep two copies around.

BTW I'm curious: in practice I never see this pattern, but I mostly focus on variational inference. What kinds of models do you see where this pattern comes up?

**Comment by iffsid** [comment]
This comes up particularly for variational inference :slightly_smiling_face: when you want a better tight estimator ([2] in the main post) than the standard IWAE estimator---because there are issues with SNR with increasing numbers of samples from the proposal/guide distribution [3].

I'm not sure why people don't extensively use this, because the difference can be quite stark!
For simple latent variable models (VAEs), this is easy enough (but ugly) to manage manually as in **Alternatives**, but for more complex/generic models, I'm baulking a bit. 

Especially for something like a Pyro model, this would effectlvely need two separate traces of the guide program, one with extra augmentation to `detach`  the parameters for the corresponding distribtution before each `pyro.sample` call. 

Maybe this could be some sort of decorator to reduce code duplication, or perhaps one could even overload the `replay` messenger to detach the sites on traces, but unsure of how practical that would be. I'd love to hear what you think might be a solution for Pyro models; happy to take the discussion offline if more appropriate somewhere else. :slightly_smiling_face: 

[3]. [Tighter Variational Bounds are Not Necessarily Better](https://arxiv.org/abs/1802.04537)

**Comment by fritzo** [comment]
cc @martinjankowiak @eb8680 @neerajprad

**Comment by iffsid** [comment]
Just for completeness, here's a simple repo showing off the use of the DReG estimator for MNIST: [iffsid/DReG-PyTorch](https://github.com/iffsid/DReG-PyTorch}. It employs an explicit `detach` on the parameters of the guide for the scoring, as in **Alternatives**.

The only operative differences from the parent repo ([lxuechen/DReG-PyTorch](https://github.com/lxuechen/DReG-PyTorch)) are:
 * a fix for detaching variables (https://github.com/iffsid/DReG-PyTorch/commit/f4e549d7c54b56c78e18b35f01912e262c278b11)
 * a way to compute the whole objective in a single pass; as opposed to two freeze-and-run steps as in the parent repo (https://github.com/iffsid/DReG-PyTorch/commit/ff3c10a6219830ea80153431d89a8854915c682e)

**Comment by neerajprad** [comment]
> api idea: maybe intro a method detach() on distributions and call it dist.detach().rsample(...). or is it too ambiguous?

> I agree that's a reasonable api, but until we can generically access parameters, .detach would require bespoke implementations for each class, which seems prohibitively complex.

@fritzo - It might be possible to use `Distribution.arg_constraints` to provide a generic method that keeps the lazy params intact, similar to what we use for constraints checking, though I may be missing some corner cases. This relies on whether all params used by distributions populate `arg_constraints`. We should probably provide a `requires_grad` method instead which sets the `requires_grad` attribute so that we can reset the `requires_grad` flag later.

```python
def detach(self):
  for param in self.arg_constraints:
    if param not in self.__dict__:  # ignore lazy params
      continue 
    p = getattr(self, param)
    p.detach_()
```

**Comment by fritzo** [comment]
@iffsid @neerajprad as a near term solution we could exploit the `.expand()` method as proposed in https://github.com/probtorch/pytorch/pull/152

@neerajprad a fully generic `.detach()` method is tricky because some distributions cache intermediate tensors, e.g. the `._unbroadcasted_scale_tril` of `MultivariateNormal`.

**Comment by fritzo** [comment]
@neerajprad what do you think of this solution. If this is too hacky for PyTorch we could add it to Pyro instead.
```py
class Distribution(object):
    ...
    def detach(self):
        new = self._get_checked_instance(type(self), _instance)
        for key, value in new.__dict__.items():
            if hasattr(value, "detach"):
                setattr(new, key, value.detach())
        super(type(self), new).__init__(self.batch_shape, self.event_shape, validate_args=False)
        new._validate_args = self._validate_args
        return new

```

**Comment by neerajprad** [comment]
It seems that `new` will not contain any distribution parameters, but I think we can use something like below? This still seems a bit hacky, e.g. if users have their own distribution implementations where they are storing parameters in some collection type, this won't work. We'll also need separate wrappers for TransformedDistribution like above. It might be something we can put into Pyro first, and have a simplified implementation in PyTorch once https://github.com/pytorch/pytorch/issues/25932 is resolved.

```python
    def detach(self):
        new = self.expand([])
        for key, value in new.__dict__.items():
            if hasattr(value, "detach"):
                setattr(new, key, value.detach())
        return new
```

**Comment by fritzo** [comment]
@iffsid another workaround to implement this generically is to cancel unwanted gradients:
```py
def dlog_prob(self, value):
    lp_full = self.log_prob(value)
    lp_part = self.log_prob(value.detach())
    return lp_full + (lp_part.detach() - lp_part)
```
Of course this is inefficient and not very NAN safe, but it should already work for prototyping.

**Comment by iffsid** [comment]
@fritzo thanks; that's cute. :)

Realistically, for simple models, just creating a new instance using detached params should do.
The more complex cases are probably better first explored in pyro, with only changes that are really necessary pushed upstream to pytorch.
